packages/kit/src/runtime/client/singletons.js

```diff
@@ -1,7 +1,7 @@
-/** @type {any} */
+/** @type {import('./router').Router} */
```

Comment:
i could definitely be mistaken but i think we *might* want to leave this as `any` so that the types are 'correct' in the bundled files

Reply:
We're not generating the types yet, are we? It's kind of hard to say what this should be without the rest of the setup. If we don't ship a `.d.ts` then are we forcing all of our users to have `allowJs: true` and `maxNodeModuleJsDepth: 1` which we've already found out is problematic

Reply:
The built files live in `$app`, which is just an alias, so `maxNodeModuleJsDepth: 1` doesn't apply. I didn't think you did need a `.d.ts` if the _importing_ file is being typechecked? E.g. if I add this to a project that has a tsconfig.json that only contains this...

```json
{
  "compilerOptions": {
    "paths": {
      "$app/*": [".svelte/dev/runtime/app/*", ".svelte/build/runtime/app/*"]
    }
  }
}
```

...I get this in a Svelte component:

![image](https://user-images.githubusercontent.com/1162160/110883120-f74ff700-82b0-11eb-9543-a4ffb40c3d23.png)

Reply:
Ah, right. I get a bit lost with the copying sometimes. It's hard to keep track of what's where and at what times when I'm so used to everything being in `node_modules`

I'm afraid I'm more confused as to how this should work. Basically all the type imports I see in `.svelte/build/runtime/internal` are broken. E.g. I see `import(./types')`, `import(../types')`, and `import(../../types')` none of which exist. Also, even without the change being made here we're referencing `Renderer` in a way that doesn't appear to work: https://github.com/sveltejs/kit/blob/master/packages/kit/src/runtime/client/router.js#L35

Reply:
The disadvantage of this kind of path alias is that it will fail until you built/ran the project once so that these folders exist. This could mean that the project isn't build because it has type errors which exist because the project hasn't been build - deadlock.

Reply:
> Basically all the type imports I see in `.svelte/build/runtime/internal` are broken

The files that are exposed to the developer (`env.js`, `navigation.js`, `paths.js` and `stores.js`) all have simple types, and with the TS configurations I've tried no errors are surfaced for the types that are missing/incorrect in non-developer-facing `internal`/`client` files. What errors are you seeing? It's possible that we'll need to do something tricky like turning everything in those internal files into an explicit `any`, but we _definitely_ don't need to make `router`/`renderer` types available at that point because the developer never interacts with them directly.

> The disadvantage of this kind of path alias is that it will fail until you built/ran the project once so that these folders exist. This could mean that the project isn't build because it has type errors which exist because the project hasn't been build - deadlock.

Not so. As soon as you run `npm run dev` or `npm run build` the `$app` files are copied over before TypeScript gets near the files that import them. As long as you have the correct `"paths"` in your tsconfig.json (which we need to add to the project template, either directly or via an `"extends"`, preferably the latter for the sake of being able to change our minds later about where files live), the only time you should experience any broken types are before you first run `npm run dev`. If we were sufficiently motivated we could solve even that, by copying assets over when a project is first created by `npm init`.

Reply:
I'm not sure I get your original point. In your last comment it sounds like you're saying we only need to worry about the types for stuff that's in `runtime/app` (i.e. `env.js`, `navigation.js`, `paths.js` and `stores.js`). So if that's the case, then what's the worry about what the type is in `internal/singletons.js`?

Reply:
Just learned about a neat typescript feature: Annotate types you don't want in your d.ts files with `@internal` and in your tsconfig.json do `"stripInternal": true`. https://www.typescriptlang.org/tsconfig#stripInternal
Problem might be that now the d.ts file contains errors because other types rely on the now gone type. For more advanced cases it might also be worth into looking at the linked api-extractor package https://api-extractor.com/

Reply:
> So if that's the case, then what's the worry about what the type is in `internal/singletons.js`?

ah — yeah, you're right, that _is_ unnecessary